SAMv1.tex

```diff
@@ -304,6 +307,25 @@ \subsection{The header section}
 \end{longtable}
 \end{center}
 
+\subsubsection{Defined sub-sort terms}\label{sec:sub-sort}
+
+While sub-sort permits implementation dependent sorting, some terms are defined to have specific meanings.
+
+\begin{description}
+\item[lexicographical] sort order is defined as a character based dictionary sort with the character order as defined by the POSIX C locale.
+  For example ``abc'', ``abc17'', ``abc5'', ``abc59'' and ``abcd'' are in lexicographical order.
+
+\item[natural] sort order is an ordering of strings in the lexicographical order, except that multi-digit natural numbers (non-negative integers including leading zeros) are ordered as a single character with value equal to the numerical value, slotting into the collation order in the same location as single-digit numbers.  Numeric ties caused through leading zeros are broken with more-zeroes coming before fewer-zeros.  Apparent negative and fractional values are not treated as part of the number, with `\texttt{-}' and `\texttt{.}' being interpreted as all other single characters for collation purposes.
+  For example, ``abc'', ``abc+5'', ``abc-5'', ``abc.d'', ``abc03'', ``abc5'', ``abc008'', ``abc08'', ``abc8'', ``abc17'', ``abc17.+'', ``abc17.2'', ``abc17.d'', ``abc59'' and ``abcd'' are in natural order.
+
+\item[duplicate] is an unspecified order that ensures that reads that would be considered part of a ``duplicate set'' (i.e. that at most one of them will \emph{not} have bit 0x400 set if duplicate-marked) are collated together.
+  This will depend on the algorithm by which duplication-marking is performed, and for example, whether UMI and Library are taken into consideration.
```

Comment:
you are correct...but technically, so is queryname sort-order...

Reply:
Not really. `GO:query` implies much less than `SO:queryname`.

For `SO:queryname` there's an obvious lexicographic sort ordering based on QNAME that is being selected, even if implementations disagree on details.

This description of `duplicate` explains the semantic intention for determining equivalence of reads, but gives no hints about what ordering between the collated equivalence groups might look like.

Reply:
You are right. I'll suggest a tightening of the duplicate order.

Reply:
Agreed it's more GO and SO, but we have sub-sort tag not sub-group so it's natural it goes there unless we wish to expand even more on our tags.

I think it just needs clarifying this is a partial sort, so more of a group, but placed in sub-sort for simplicity.

Reply:
What value of _sort-order_ would this partner with? i.e., what is `XX` in `SO:XX SS:XX:duplicate`?

I'm not suggesting inventing a sub-group field. I'm saying this (as described) should just be `GO:duplicate`.

Reply:
I propose defining SO as _sort-order:sub-sort_ or _group-order:sub-group_.  The two should be unambiguous, given we don't have both defined at the same time (can we?) and the keywords don't clash either.

Reply:
Okay, what value of _group-order_ would this (i.e., `duplicate`) partner with?

Reply:
PS.  Argh!  Why did none of us spot GO:query vs SO:queryname at the time?  Bah humbug...  (I include myself in the list of failings for that one.)

Reply:
> Why did none of us spot GO:query vs SO:queryname at the time

Because that was back in the 1000 genomes Pages word processor document days, and most of the current “us” were not involved at the time, nor had any input or visibility into what G1K were doing.

Reply:
Actually I take back the stuff above.  I misunderstood the argument I think, as maybe have you John.

The duplicate sub-sort is still *sorted* on the primary key, so it has nothing to do with group order.  It is strictly a sub-field of sort, not group.  Why does GO enter in this at all?

Reply:
I considered that possibility, but that's not what the proposed text says. Or, at least…

It might be useful to have an ordering that is e.g. “coordinate sorted, except that records that are duplicates are grouped together”. However such a file is no longer sorted by coordinate!

So if we wanted this to mean that, then we have to add a bunch of text explaining that we don't mean that files have to be **entirely** sorted. I don't think we want to go down that road…

Reply:
I think we need @yfarjoun to explain the intended purpose.  I'm now confused.

We obviously interpreted "unspecified order" differently.  @yfarjoun, does that mean unsorted primary order, or that duplicate sub-sort doesn't have any requirements as to what the primary sort is?  The way I read it was more that in eg SS:queryname:duplicate we're explicitly stating only a maximum of one field will be non-duplicate.

Can you give an explanation of what problem you're solving with this that isn't, for example, solved by GO:query or SO:queryname already?